include/ensmallen_bits/nelder_mead/nelder_mead.hpp

```diff
@@ -0,0 +1,234 @@
+ *
+ * For more information, please refer to:
+ *
+ * @code
+ * @article{Nelder1965ASM,
+ *   author  = {J. Nelder and R. Mead},
+ *   title   = {A Simplex Method for Function Minimization},
+ *   year    = {1965},
+ *   journal = {Comput. J.},
+ *   volume  = {7},
+ *   pages   = {308-313}
+ * }
+ * @endcode
+ *
+ * @code
+ * @article{Fuchang2012,
+ *   author    = {Gao, Fuchang and Han, Lixing},
+ *   title     = {Implementing the Nelder-Mead Simplex Algorithm with Adaptive
+ *                Parameters},
+ *   year      = {2012},
+ *   publisher = {Kluwer Academic Publishers},
+ *   volume    = {51},
+ *   number    = {1},
+ *   journal   = {Comput. Optim. Appl.},
+ *   month     = jan,
+ *   pages     = {259–277}
+ * }
+ * @endcode
+ *
+ * NelderMead can optimize arbitrary functions.  For more details, see the
+ * documentation on function types included with this distribution or on the
+ * ensmallen website.
+ *
+ * @tparam SimplexerType Simplex policy used by Nelder Mead to construct the
+ * initial simplex.
+ */
+template<typename SimplexerType = AffineSimplexer>
+class NelderMeadType
+{
+ public:
+  /**
+   * Construct the NelderMead optimizer with the default adaptive parameters
+   * scheme as used in "Implementing the Nelder-Mead simplex algorithm with
+   * adaptive parameters" by F. Gao and L. Han (2010). These are based on the
+   * dimensionality of the problem, and are given by
+   *
+   * alpha = 1
+   * beta = 1 + 2 / n
+   * gamma = 0.75 - 0.5 / n
+   * delta = 1 - 1 / n
+   *
+   * where n is the dimensionality of the problem.
+   *
+   * The defaults here are not necessarily good for the given problem, so it is
+   * suggested that the values used be tailored to the task at hand.  The
+   * maximum number of iterations refers to the maximum number of points that
+   * are processed (i.e., one iteration equals one point; one iteration does not
+   * equal one pass over the dataset).
+   *
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *     limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   * @param simplexer The simplex policy used to construct the initial simplex.
+   */
+  NelderMeadType(const size_t maxIterations = 100000,
+                 const double tolerance = 1e-15,
+                 const SimplexerType& simplexer = SimplexerType());
+  /**
+   * Construct the NelderMead optimizer with the given parameters. The defaults
+   * here are not necessarily good for the given problem, so it is suggested
+   * that the values used be tailored to the task at hand.  The maximum number
+   * of iterations refers to the maximum number of points that are processed
+   * (i.e., one iteration equals one point; one iteration does not equal one
+   * pass over the dataset).
+   *
+   * @param alpha The reflection parameter.
+   * @param beta The expansion parameter.
+   * @param gamma The contraction parameter.
+   * @param delta The shrink step parameter.
```

Comment:
It might be pointing out that the ANMS paper uses different notation here than the Wikipedia description of Nelder-Mead; on Wikipedia, gamma is the expansion parameter, and in the ANMS paper, beta is...  I don't think we need to change the implementation, but we should probably point that out. :+1: